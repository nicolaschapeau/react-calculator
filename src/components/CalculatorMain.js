import React from 'react'
import CalculatorCore from '../utils/CalculatorCore'
import CalculatorScreen from './CalculatorScreen'
import CalculatorPad from './CalculatorPad'



// Component
class CalculatorMain extends React.Component {
    // State
    state = {
        calculator: new CalculatorCore(),
        expression: '',
        message: '',
        value: '',
        active: false,
    }



    // On pad click
    onPadClick = (arg) => {
        // Set new expression
        let expression = this.state.expression + `${arg}`

        // Update state
        this.setState({
            expression,
            value: '',
            message: '',
            active: false
        })
    }


    
    // On submit of input
    onSubmit = () => {
        // Get response
        let response = this.state.calculator.calculateExpression(this.state.expression)

        // Success
        if (response.success) {
            this.setState({
                expression: response.expression,
                message: response.message,
                value: response.value,
                active: true
            })
        // Error
        } else {
            this.setState({
                message: response.message,
                value: '',
                active: true
            })
        }
    }



    // On pad click
    onClear = () => {
        // Update state
        this.setState({
            expression: '',
            value: '',
            active: false
        })
    }



    // Render
    render () {
        return (
            <React.Fragment>
                <div className="calculator--display--container">
                    <CalculatorScreen
                        expression={this.state.expression}
                        message={this.state.message} 
                        value={this.state.value}
                        active={this.state.active} 
                    />
                </div>
                <CalculatorPad
                    onSubmit={(e) => this.onSubmit(e)}
                    onPadClick={this.onPadClick}
                    onClear={this.onClear}
                />
            </React.Fragment>
        )
    }
}



// Export
export default CalculatorMain