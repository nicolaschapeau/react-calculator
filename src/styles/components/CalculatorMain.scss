.calculator--display--container {
    position: relative;
    margin: 15px 0px 15px 0px;
    width: 90%;
    display: flex;
    justify-content: flex-start;
    align-items: flex-end;
    flex-direction: column;
    border-radius: 5px;
    box-sizing: border-box;
    background: white;


    .calculator-displayer {
        height: 80px;
        margin-bottom: 7px;
        width: 100%;
        display: flex;
        justify-content: flex-start;
        align-items: flex-end;

        p {
            font-family: Arial, Helvetica, sans-serif;
            width: 100%;
            margin: 0;
            padding-right: 10px;
            text-align: right;
            font-size: 24px;
            font-weight: bold;
            -webkit-transition: 0.2s;
            transition: 0.2s;
        }
    }
}


.calculator-pad-container {
    width: 90%;
    display: flex;
    justify-content: flex-start;
    align-items: flex-start;
    flex-direction: row;
    flex-wrap: wrap;
    margin-bottom: 20px;

    .calculator-pad-item {
        cursor: pointer;
        margin: 6px 2% 6px 2%;
        width: 21%;
        height: 50px;
        color: white;
        border-radius: 5px;
        text-align: center;
        display: flex;
        justify-content: center;
        align-items: center;

        &:hover {
            margin: 7px 2% 5px 2%;
        }

        &:active {
            margin: 8px 2% 4px 2%;
            -webkit-box-shadow: 0px 0px 0px 0px white !important;
            -moz-box-shadow: 0px 0px 0px 0px white !important;
            box-shadow: 0px 0px 0px 0px white !important;
        }

        p {
            font-family: Arial, Helvetica, sans-serif;
            margin: 0;
            font-size: 21px;
        }
    }


    .calculator-pad-item.number {
        background: #37474F;
        -webkit-box-shadow: 0px 4px 0px 0px #263238;
        -moz-box-shadow: 0px 4px 0px 0px #263238;
        box-shadow: 0px 4px 0px 0px #263238;

        &:hover {
            -webkit-box-shadow: 0px 3px 0px 0px #263238;
            -moz-box-shadow: 0px 3px 0px 0px #263238;
            box-shadow: 0px 3px 0px 0px #263238;
        }
    }

    .calculator-pad-item.symbol {
        background: #607D8B;
        -webkit-box-shadow: 0px 4px 0px 0px #546E7A;
        -moz-box-shadow: 0px 4px 0px 0px #546E7A;
        box-shadow: 0px 4px 0px 0px #546E7A;

        &:hover {
        -webkit-box-shadow: 0px 3px 0px 0px #546E7A;
        -moz-box-shadow: 0px 3px 0px 0px #546E7A;
        box-shadow: 0px 3px 0px 0px #546E7A;
        }
    }

    .calculator-pad-item.result {
        background: #FF9800;
        -webkit-box-shadow: 0px 4px 0px 0px #FB8C00;
        -moz-box-shadow: 0px 4px 0px 0px #FB8C00;
        box-shadow: 0px 4px 0px 0px #FB8C00;

        &:hover {
            -webkit-box-shadow: 0px 3px 0px 0px #FB8C00;
            -moz-box-shadow: 0px 3px 0px 0px #FB8C00;
            box-shadow: 0px 3px 0px 0px #FB8C00;         
        }
    }

    .calculator-pad-item.clear {
        background: #F44336;
        -webkit-box-shadow: 0px 4px 0px 0px #D32F2F;
        -moz-box-shadow: 0px 4px 0px 0px #D32F2F;
        box-shadow: 0px 4px 0px 0px #D32F2F;

        &:hover {
            -webkit-box-shadow: 0px 3px 0px 0px #D32F2F;
            -moz-box-shadow: 0px 3px 0px 0px #D32F2F;
            box-shadow: 0px 3px 0px 0px #D32F2F;         
        }
    }
}