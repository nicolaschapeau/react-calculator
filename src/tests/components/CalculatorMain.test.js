import React from 'react';
import { shallow, mount } from 'enzyme';
import CalculatorMain from '../../components/CalculatorMain';
import toJSON from 'enzyme-to-json'


test('Should render CalculatorMain correctly', () => {
    const wrapper = shallow(<CalculatorMain />)

    expect(toJSON(wrapper)).toMatchSnapshot()
})



// onSubmit
test('Should execute onSubmit method and mutate the state properly', () => {
    const wrapper = mount(<CalculatorMain />)

    wrapper.setState({ expression: "3+2" })
    wrapper.instance().onSubmit()

    expect(wrapper.state('value')).toEqual(5)
})



// onPadClick
test('Should execute onPadClick method and mutate the state properly', () => {
    const wrapper = mount(<CalculatorMain />)

    wrapper.setState({ expression: "3+2" })
    wrapper.instance().onPadClick('+5')

    expect(wrapper.state('expression')).toEqual('3+2+5')
})



// onAns
test('Should execute onAns method and mutate the state properly', () => {
    const wrapper = mount(<CalculatorMain />)

    wrapper.setState({ ans: "45", value: '', expression: "25+20+" })
    wrapper.instance().onAns()

    expect(wrapper.state('expression')).toEqual('25+20+45')
})



// onDelete
test('Should execute onDelete method and mutate the state properly', () => {
    const wrapper = mount(<CalculatorMain />)

    wrapper.setState({ expression: "25+20+" })
    wrapper.instance().onDelete()

    expect(wrapper.state('expression')).toEqual('25+20')
})



// onClear
test('Should execute onClear method and mutate the state properly', () => {
    const wrapper = mount(<CalculatorMain />)

    wrapper.setState({ expression: "25+20+" })
    wrapper.instance().onClear()

    expect(wrapper.state('expression')).toEqual('')
})
