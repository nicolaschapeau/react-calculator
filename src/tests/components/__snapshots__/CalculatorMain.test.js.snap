// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render CalculatorMain correctly 1`] = `
<Fragment>
  <div
    className="calculator--display--container"
  >
    <CalculatorScreen
      active={false}
      expression=""
      message=""
      value=""
    />
  </div>
  <CalculatorPad
    onAns={[Function]}
    onClear={[Function]}
    onDelete={[Function]}
    onPadClick={[Function]}
    onSubmit={[Function]}
  />
</Fragment>
`;
